torch/csrc/jit/runtime/interpreter.cpp

```diff
@@ -1017,6 +1034,22 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
   }
 
  private:
+  struct WarnedNodes {
+   public:
+    // Inserts idx into warned_nodes_, returns a boolean indicates whether
+    // insertion actually happened (idx wasn't originally in the set).
+    bool insert(int32_t idx) {
+      std::unique_lock<std::mutex> lock(mutex_);
+      return warned_nodes_.insert(idx).second;
+    }
+
+   private:
+    std::mutex mutex_;
+    std::unordered_set<int32_t> warned_nodes_;
+  };
+
+  WarnedNodes warned_nodes_;
```

Comment:
This is actually not we want - multiple invocations of the same function should still warn only once (like python does). Sorry for missing it in the review. You probably need to move it to Code level or something like this

Reply:
Another suggestion - instead of unordered_set, just create a vector<atomic<bool>> inside Code and use that to modify the index directly. I think this way you don't need locking. Or you can still have unordered_map, but pre-populate it with all possible indices beforehand so that you don't need the lock

Reply:
Thanks for the suggestion, I wonder if what python does is actually desired. Think about following use case:

```
@torch.jit.script
def issue_warning_wrong_dtype(dtype: str):
    warnings.warn("Incorrect data type " + dtype)
```

It may be called in various callsites, and they are all useful. IMHO, it is slightly emit warning for *each unique callsite*. Number of warnings would still not be too spammy because there likely won't be that many callsites to the extend that they feel spammy to users. What do you think?

I am still investigating the issue reported internally, which I feel is a multi-threading issues similar to #46684. Will update here once finding out more.

Reply:
Well, ideally we should respect `stacklevel` argument, like Python does. I.e. in you `issue_warning_wrong_dtype` example, it'd be `stacklevel=2` or something. However, that's much harder to implement and frankly I don't think it's worth the effort as warnings in TorchScript are kind of a niche use case in general.

Fixing the spammy warning is important though, because the typical case is inference, where model is called many times in a loop and each of them creates independent InterpreterState. So the current implementation sadly will log on each invocation spamming the logs. (that's pretty much the internal issue you're referring to).

Reply:
I thought about this problem a lot but couldn't think of a clean way to implement the "warn once" behavior, which requires maintaining a global state that is above ScriptModule. This sort of goes against TorchScript's philosophy of keep states of ScriptModule local to itself. 

Here is an alternative solution: Given what we talked about, it sounds like it is OK (or even preferred) to have zero warnings emitted during inference. If that's the case, we can expose a control knob that predictor can toggle to disable Warnings entirely inside TorchScript module. What do you think?

Reply:
Yeah, we can have a global knob :)

In general, I don't think that having this state across ScriptModule invocation is necessarily that bad. It's a very narrow case of state and it matches python's `warnings` module behavior.

> requires maintaining a global state that is above ScriptModule

I wonder whether we can just add a bit per original Warning IR Node somewhere. That'd be the closest match to what python does and also keep state "local" to ScriptModule that owns the original graph.

Reply:
> I wonder whether we can just add a bit per original Warning IR Node somewhere. That'd be the closest match to what python does and also keep state "local" to ScriptModule that owns the original graph.

Based on my rough understanding, that bit per Warning node would have to be in predictor, which runs inference calls in a for loop. In that case, implementation detail of TorchScript kind of leaked. Let me know what you think.